### DISCUSSION TITLE
passing input from top level drawables to parent

### DISCUSSION THREAD

**Discussion by approachcircle** [ppy/osu-framework#5393] [Q&A] [answered]
i was wondering if there is a way to disallow a drawable from accepting input directly, and instead passing the input up to the drawable's parent to be dealt with instead using the parent's event handlers (OnClick, OnHover). i originally thought that if a top level drawable did not accept positional or non positional input, then the input would just be passed up the hierarchy, but it appears that's not the case. 

apologies if i'm demonstrating terrible practice or do not have a good understanding of osu!framework, i'm very new to using it.

**Comment by peppy** [comment, marked as answer]
can you provide a code sample? i'm having trouble understanding what you are doing/seeing.

**Reply by approachcircle** [reply]
something along the lines of this

    public class ButtonContainer : ClickableContainer
    {
        [BackgroundDependencyLoader]
        private void load()
        {
            AddInternal(new UselessButton());
        }

        protected override bool OnClick(ClickEvent e)
        {
            /** we want the ButtonContainer's OnClick method to be called
             * instead of its child (UselessButton) handling the input,
             * by somehow passing the input received by UselessButton up to its
             * parent (ButtonContainer). this may be impossible or impractical, and if so,
             * i'm fine with just finding a way around it.
            **/
            return true;
        }
    }

    public class UselessButton : Container
    {
        [BackgroundDependencyLoader]
        private void load()
        {
            AddInternal(new Box());
        }
    }

**Reply by frenzibyte** [reply]
The parent will receive a mouse click as long as it's not handled by the child drawable, so if `UselessButton` doesn't handle input, then `ButtonContainer` will receive the event and can handle it.

Also, since you're using `ClickableContainer`, you would assign the callback to `Action`.

**Reply by approachcircle** [reply]
i have a `ClickableContainer` that i have now assigned the `OnClick` callback to `Action` with, and it has a child `Container` that does not contain any event handlers at all, but for some reason input is still not being handled by the `ClickableContainer`, and is instead being consumed by the child `Container`. the `ClickableContainer` has other callbacks such as `OnHover`, `OnMouseDown` that are also not being fired

**Reply by smoogipoo** [reply]
If your case is that you don't want children to handle any input whatsoever, then you could override `ShouldBeConsideredForInput(Drawable child) => false`.

**Reply by approachcircle** [reply]
i overrode `ShouldBeConsideredForInput(Drawable child) => false` on the parent, and still it is not receiving any events. i have also just realised that the child `Container` to the parent `ClickableContainer` in my project actually isn't receiving input in the first place, because when it is clicked, nothing shows in the log stating that anything has actually handled input, so i have no idea where the input is going.

i tried replacing the `ClickableContainer`'s child `Container` with just a `Box`, but still nothing is receiving input.

**Reply by smoogipoo** [reply]
Please provide a complete test project or test scene. The code you've provided above is incomplete since you mention having callbacks that aren't listed.

Based on just the code above, the `Box` won't receive input and won't indicate input was received, because it doesn't override any of the handlers.

**Reply by approachcircle** [reply]
i have fixed the issue now, and the reason my ClickableContainer was never actually receiving input is because i didn't scale or move the parent container with it's children properly, so the position of it would never be in the same place as it's children, nor the same size, meaning it physically couldn't handle any events.

all callbacks are now being fired correctly, i just need to work on polishing the relative position of the parent container and its children, which shouldn't be a problem.
